recap/src/Imp/Eval.hs

```diff
@@ -17,14 +17,53 @@ data Error = UndefinedVar String | ParsingErr String | DivByZero deriving (Show)
 -- Evaluation monad
 type EvalM = StateT VarMap (ExceptT Error IO)
 
+throw = lift . throwE
+
 evalExpr :: Expr -> EvalM Int
-evalExpr = undefined 
+evalExpr (Var x) = do
+  varMap <- get
+  case M.lookup x varMap of
+    Just x -> return x
+    Nothing -> throw (UndefinedVar x)
+evalExpr (Const n) = return n
+evalExpr (BinOp op expr1 expr2) = do
+  e1 <- evalExpr expr1
+  e2 <- evalExpr expr2
+  case op of
+    Plus -> return (e1 + e2)
+    Minus -> return (e1 - e2)
+    Mult -> return (e1 * e2)
+    Div -> case e2 of
+      0 -> throw DivByZero
+      _ -> return (e1 `div` e2)
 
 evalCom :: Com -> EvalM ()
-evalCom = undefined 
+evalCom (Assign name expr) = do
+  value <- evalExpr expr
+  modify $ M.insert name value
+evalCom (Read name) = do
+  value <- lift $ lift getLine
+  case readMaybe value of
+    Just value -> evalCom $ Assign name $ Const value
```

Comment:
It's also totally fine to do the straightforward insertion. The potential for error introduction is the same in both cases.